src/core/net/ip4_address.hpp

```diff
@@ -101,12 +126,75 @@ class Address : public Equatable<Address>, public Clearable<Address>
      *
      */
     InfoString ToString(void) const;
-
-private:
-    uint8_t mBytes[kSize];
 } OT_TOOL_PACKED_END;
 
+OT_TOOL_PACKED_BEGIN
+class Cidr : public otIp4Cidr, public Unequatable<Cidr>, public Clearable<Address>
+{
+public:
+    static constexpr uint16_t kCidrSuffixSize = 3; ///< Suffix to represent CIDR (/dd).
+
+    /**
+     * This type defines the fixed-length `String` object returned from `ToString()`.
+     *
+     */
+    typedef String<Address::kAddressStringSize + kCidrSuffixSize> InfoString;
+
+    /**
+     * This method converts the IPv4 address to a string.
+     *
+     * The string format uses quad-dotted notation of four bytes in the address (e.g., "127.0.0.1").
+     *
+     * @returns An `InfoString` representing the IPv4 address.
+     *
+     */
+    InfoString ToString(void) const;
+
+    /**
+     * This method returns the host mask (bitwise not of the subnet mask) of the CIDR.
+     *
+     * @returns An uint32 for the host mask, in network byte order.
+     */
+    inline uint32_t HostMask() const { return HostSwap32((uint32_t(1) << uint32_t(32 - mLength)) - 1); }
+
+    /**
+     * This method returns the subnet mask of the CIDR.
+     *
+     * @returns An uint32 for the subnet mask, in network byte order.
+     */
+    inline uint32_t SubnetMask() const { return ~HostMask(); }
+
+    /**
+     * This method gets the prefix as a pointer to a byte array.
+     *
+     * @returns A pointer to a byte array containing the Prefix.
+     *
+     */
+    const uint8_t *GetBytes(void) const { return mAddress.mFields.m8; }
+
+    /**
+     * This method overloads operator `==` to evaluate whether or not two prefixes are equal.
+     *
+     * @param[in]  aOther  The other prefix to compare with.
+     *
+     * @retval TRUE   If the two prefixes are equal.
+     * @retval FALSE  If the two prefixes are not equal.
+     *
+     */
+    bool operator==(const Cidr &aOther) const;
+
+    /**
+     * This method returns an IPv4 address within the CIDR block.
+     *
+     * @param[in] aHost The host bits in of the IPv4 address in host byte order. The aHost will be masked by host mask.
+     */
+    Address Host(uint32_t aHost) const;
```

Comment:
While this works and is correct, in general, I would suggest avoid returning objects from methods/functions when possible. Instead we can pass a reference to the object to be set/populated.

RVO  "Return Value Optimization" (https://en.cppreference.com/w/cpp/language/copy_elision) should in general optimize this, but in many cases it is optional (up to toolchain whether to optimize this).  

OT is an embedded project and is often built with platform specific embedded toolchain/compiler which may not support this. It is better to write the code in way that it is efficient on all toolchains.